subsys/net/ip/net_shell.c

```diff
@@ -3926,6 +3926,27 @@ static void tcp_sent_cb(struct net_context *context,
 {
 	PR_SHELL(tcp_shell, "Message sent\n");
 }
+
+static void tcp_recv_cb(struct net_context *context,
+			struct net_pkt *pkt,
+			int status,
+			void *user_data)
+{
+	if (status < 0) {
+		PR_SHELL(tcp_shell, "Message receive error: %d\n", status);
+		return;
+	}
+
+	if (!pkt) {
+		PR_SHELL(tcp_shell, "Message received: EOF\n");
```

Comment:
I recommend to use `printk` here. If I understand it correctly this is callback function which will be called outside of command context. This can lead to incorrect text display on the screen.

Reply:
Well, I don't know about the underlying of PR_SHELL, this function is indeed called outside of the command handler, but is still synced with the handler.
I mean the command handler calls net_context_recv() which is a blocking call, the recv callback can only occur during the net_context_recv duration.

Reply:
I guess what @jarz-nordic means is that if some callback is called from other context than shell per se (and that's what happens with net_context_recv()), then it shouldn't use shell funcs. Worst case, that can lead to some reentrancy problems, etc.

Reply:
The tcp_shell variable is set when user does tcp connect command, so no need to use printk() here.